### ISSUE TITLE
Very slow pre-processing of images before the training stage.

### ISSUE THREAD

**Issue by evgenykam** [alexjc/neural-enhance#131]
Hey thanks for the great code!

The only problem that I have yet found is a very slow  image preprocessing before the training stage.

So, I'm trying to "train the model using an adversarial setup based on" SRGAN for the Celeb-A dataset (i.e. 218x178, "Align & Crop images"), so I changed the default "batch-shape" to 96, as mentioned in the original SRGAN article).

python3.4 enhance.py \
         --train "data/*.png" --model custom \
         --epochs=250 --batch-shape=96 --device=gpu0 \
         --generator-downscale=1 --generator-upscale=2 \
         --perceptual-layer=conv5_2 --smoothness-weight=2e4 --adversary-weight=1e3 \
         --generator-start=5 --discriminator-start=0 --adversarial-start=5 \
         --discriminator-size=64

But, when I run "enchance.py", I have the following problem:

Preprocessing of images  before the training phase is performed only with the use of one processor core. For example, using (E5-2670 v3) for 100 Celeb-A images, this takes about 30-40 minutes, then the training stage begins, which performs quite quickly. So, for training using all 162K training images, I'm still waiting for the end of the preprocessing stage (I waited for more than 24 hours and killed the process). Is there a way to fix the above issue or what do I do to fix it?

Thank you in advance!

**Comment by evgenykam** [comment]
So far, I've found a patch for solving this problem. If I download a file with a model, everything works fine. To create the training model, added a dummy process, then stopped it with Cntrl-C,  then bash starts the normal process automatically.

#!/bin/sh

export OPEN_IMAGES_PATH=~/neural-enhance-master/data/small
 
# dummy process
  python3.4 enhance.py \
           --train "$OPEN_IMAGES_PATH/*.png" --type photo --model default \
           --epochs=1 --batch-shape=96 --device=gpu0 \
           --generator-downscale=1 --generator-upscale=2 \
           --perceptual-layer=conv5_2 --smoothness-weight=2e4 --adversary-weight=1e3 \
           --generator-start=5 --discriminator-start=0 --adversarial-start=5 \
           --discriminator-size=64

# Stop "dummy process" with Ctrl-C after you get "tensors learned for discriminator"  


# normal process
export OPEN_IMAGES_PATH=/data/work-data-local/evgeny/open_image/train

#THEANO_FLAGS='floatX=float32,device=gpu0,lib.cnmem=0' 
python3.4 enhance.py \
    --train "$OPEN_IMAGES_PATH/*/*.jpg" --type photo --model default \
    --epochs=50 --batch-shape=256 --device=gpu0 \
    --generator-downscale=0 --generator-upscale=1 \
    --generator-blocks=8 --generator-filters=128 --generator-residual=0 \
    --perceptual-layer=conv2_2 --smoothness-weight=1e7 --adversary-weight=0.0 \
    --train-noise=1.0

python3.4 enhance.py \
    --train "$OPEN_IMAGES_PATH/*/*.jpg" --type photo --model default \
    --epochs=500 --batch-shape=240 --device=gpu0 \
    --generator-downscale=0 --generator-upscale=1 \
    --perceptual-layer=conv5_2 --smoothness-weight=5e3 --adversary-weight=5e1 \
    --generator-start=10 --discriminator-start=0 --adversarial-star
    --discriminator-size=64 \
    --train-noise=1.0

**Comment by sdlovecraft** [comment]
I am also having this problem, I have 110k 200x200 images that I wish to train an adversarial upscaling model on, however it has been 2 days, and I am still in the pre-processing stage.

I am attempting to train with the command:

`python enhance.py --train "./nolikes/*.jpg" --model custom --epochs=2000 --perceptual-layer=conv5_2 --device=gpu0 --smoothness-weight=2e4 --adversary-weight=1e3 --generator-start=5 --discriminator-start=0 --adversarial-start=5 --discriminator-size=64`

@evgenykam I am unclear on you instructions, Did you mean to run the normal process and dummy process at the same time? or is the dummy process run first, cancelled and then the normal process is run? why did you mention 2 different commands in the normal process?

**Comment by sdlovecraft** [comment]
could this line be the problem @alexjc @evgenykam ? Seems like we are entering an infinite loop in preprocessing here..

https://github.com/alexjc/neural-enhance/blob/master/enhance.py#L158

**Comment by evgenykam** [comment]
@sdlovecraft 
I run all three commands in one bash script:  "dummy", then the pre-tarining stage (SRResNet training), and finally the GAN training (with conv5_2 or conv5_4).

**Comment by xycyx** [comment]
Hi, @evgenykam 
I met the same situation. I am trying to train a model based on my 256*256*10k images, but the pre-processing still takes a long time and the program has no output. I tried your dummy method but it seems still waiting for the processing. Could you tell me how long did the preprocessing take after using the dummy method?

Thank you!
![screen shot 2017-06-05 at 2 25 22 am](https://cloud.githubusercontent.com/assets/16574534/26778807/afc4b32c-4997-11e7-8b94-5754c3069b1c.png)

**Comment by wildeyevr** [comment]
I agree with samhains comment... I'm actively working on this issue... I've also got a gui in progress if anyone is interested... I can share soon...

/r

**Comment by wildeyevr** [comment]
Also not sure what OS you guys are on with this issue... but I have run this with no problem on OSX... may be a windows thing...
/r

**Comment by wildeyevr** [comment]
update... it seems this project really needs to be updated to the latest theano/lasagne api versions... I will create a branch...

/r

**Comment by wildeyevr** [comment]
hey guys... I've got the training working normally now on current theano and lasagne api... I don't have enough memory to test on the gpu... but it's training just fine on the cpu... i'll get a repo going but if you want me to send over in the meantime email me at ryan@proscans.co

**Comment by wildeyevr** [comment]
hey guys another thing... the original code seems to work fine at smaller --batch-shape values... it goes along training normally at --batch-shape=96 for instance... but hangs when I set up to 256...

**Comment by wildeyevr** [comment]
followup to my last comment... funny enough it works out of the box on ubuntu 16... however... if the batch_shape parameter is set too high... like in the examples it's 256... i get an intolerably long hang... setting it to something like 96 performs normally...

/r
